akka-actor/src/main/scala/akka/pattern/Patterns.scala

```diff
@@ -485,6 +485,44 @@ object Patterns {
       scheduler: Scheduler,
       ec: ExecutionContext): CompletionStage[T] =
     scalaRetry(() => attempt.call().toScala, attempts, delay.asScala)(ec, scheduler).toJava
+
+  /**
+   * Returns an internally retrying [[scala.concurrent.Future]].
+   * The first attempt will be made immediately, and each subsequent attempt will be made after 'delay'.
+   * A scheduler (eg context.system.scheduler) must be provided to delay each retry.
+   * You could provide a function to generate the next delay duration after first attempt,
+   * this function should never return `null`, otherwise an [[IllegalArgumentException]] will be through.
+   * If attempts are exhausted the returned future is simply the result of invoking attempt.
+   * Note that the attempt function will be invoked on the given execution context for subsequent tries and
+   * therefore must be thread safe (not touch unsafe mutable state).
+   */
+  def retry[T](
+      attempt: Callable[Future[T]],
+      attempts: Int,
+      delayFunction: akka.japi.function.Function[Int, FiniteDuration],
+      scheduler: Scheduler,
+      context: ExecutionContext): Future[T] =
+    scalaRetry(() => attempt.call, attempts, attempted ⇒ Option(delayFunction.apply(attempted)))(context, scheduler)
+
+  /**
+   * Returns an internally retrying [[java.util.concurrent.CompletionStage]].
+   * The first attempt will be made immediately, and each subsequent attempt will be made after 'delay'.
+   * A scheduler (eg context.system.scheduler) must be provided to delay each retry.
+   * You could provide a function to generate the next delay duration after first attempt,
+   * this function should never return `null`, otherwise an [[IllegalArgumentException]] will be through.
+   * If attempts are exhausted the returned future is simply the result of invoking attempt.
+   * Note that the attempt function will be invoked on the given execution context for subsequent tries and
+   * therefore must be thread safe (not touch unsafe mutable state).
+   */
+  def retry[T](
+      attempt: Callable[CompletionStage[T]],
+      attempts: Int,
+      delayFunction: akka.japi.function.Function[Int, java.time.Duration],
```

Comment:
If it is possible to stop retrying with an `Option` in Scala I think the Java API should support it as well using an `Optional`.

Reply:
Will update later